[SUREFIRE-2063] Remove tabs in forked JVM arguments

mthmulders opened the pull request:
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SUREFIRE) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
       [**SUREFIRE-2063**](https://issues.apache.org/jira/browse/SUREFIRE-2063)
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[SUREFIRE-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `SUREFIRE-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean install` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [X] You have run the integration tests successfully (`mvn -Prun-its clean install`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [ ] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [X] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Tibor17 commented:
We have too many ITs and adding such one would slow down the build more and more.
Is it possible to find very similar and existing IT with argLine and make such addons with white spaces?

mthmulders commented:
> Is it possible to find very similar and existing IT with argLine and make such addons with white spaces?

Looking into that. Strange thing: when I wrote the IT, I made sure it failed without the change in `DefaultForkConfiguration` - but now it seems that change has no effect and the test fails for another reason. **This change is definitely not mergeable yet!**

mthmulders commented:
> > Is it possible to find very similar and existing IT with argLine and make such addons with white spaces?
> 
> Looking into that. Strange thing: when I wrote the IT, I made sure it failed without the change in `DefaultForkConfiguration` - but now it seems that change has no effect and the test fails for another reason. **This change is definitely not mergeable yet!**

Of course. It's always tabs vs. spaces. Replacing spaces with tabs makes the test consistently fail again. I'll get back to this PR later this week!

Tibor17 commented:
@mthmulders 
Did you use JDK1.8 on your PC when you tested the IT?
Look at the GH CI results. All JDKs fail except 1.8.
Is this the reason?

mthmulders commented:
> Did you use JDK1.8 on your PC when you tested the IT?
> Look at the GH CI results. All JDKs fail except 1.8.
> Is this the reason?

I'll soon push an update to an existing test, dropping the need for the extra test case. Also, that test runs on all versions of Java, rather than using module flags.

Tibor17 commented:
@mthmulders What was the reason that the only Java `1.8` was working in the old IT? I am just interested.

Tibor17 commented:
@mthmulders 
I have one more question.
The developers would not be able to see the TAB because it is a new space.
If the new guy would refactor it again, then this IT would not cover the purpose that it has now.
I know it is a question how the xml parser in Maven interprets the escaping characters, but I would like to ask you for testing the escaping characters  `&#09;` on Maven 3.2.x and 3.8.x.
WDYT?

dsubelman commented:
If the test fails with escaping characters `&#09;`, you can try using `String.replaceAll(String, String)` with regex.

Instead of:

    return argLine
        .replace( "\n", " " )
        .replace( "\r", " " )
        .replace( "\t", " " );

You can do:

    return argLine.replaceAll("\\s+", " ");

mthmulders commented:
> @mthmulders What was the reason that the only Java `1.8` was working in the old IT? I am just interested.

At first, I thought (due to the original report in JIRA) that the problem only appeared when using module flags. So my test case mimicked that. But the test had a problem causing it to fail. Since the test was written to only run on Java 9 and above, it would always fail on Java 9+, but never run on Java 1.8 - hence only 1.8 was green...

mthmulders commented:
> @mthmulders I have one more question. The developers would not be able to see the TAB because it is a white space. If the new guy would refactor it again, then this IT would not cover the purpose that it has now. I know it is a question of how the xml parser interprets the escaping characters in Maven but I would like to ask you for testing with escaping characters `&#09;` on Maven 3.2.x and 3.8.x. WDYT?

That's why I added XML comments saying those lines must be tab-indented. I've just tested locally: an XML file with `&#09;` also breaks before the change in `DefaultForkConfiguration` (tested with Maven 4 SNAPSHOT, Maven 3.8.5 and Maven 3.2.5). So it seems to work equally fine for the test. Just tell me which one you prefer @Tibor17: tab characters, XML entities, or both.

mthmulders commented:
> If the test fails with escaping characters `&#09;`, you can try using `String.replaceAll(String, String)` with regex.
> 
> Instead of:
> 
> ```
> return argLine
>     .replace( "\n", " " )
>     .replace( "\r", " " )
>     .replace( "\t", " " );
> ```
> 
> You can do:
> 
> ```
> return argLine.replaceAll("\\s+", " ");
> ```

It does, but it also matches spaces (which it currently doesn't). On the other hand, it's one regex vs. three, so that could be a benefit. What do you think, @Tibor17?

dsubelman commented:
> > If the test fails with escaping characters `&#09;`, you can try using `String.replaceAll(String, String)` with regex.
> > Instead of:
> > ```
> > return argLine
> >     .replace( "\n", " " )
> >     .replace( "\r", " " )
> >     .replace( "\t", " " );
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can do:
> > ```
> > return argLine.replaceAll("\\s+", " ");
> > ```
> 
> It does, but it also matches spaces (which it currently doesn't). On the other hand, it's one regex vs. three, so that could be a benefit. What do you think, @Tibor17?

It also replaces multiple consecutive whitespaces with a single whitespace:
        
    List<String> whitespaces = List.of(
            "\n", 
            "\r",
            "\n\r",
            "\t", 
            " ",                    // single whitespace
            "  ",                   // double whitespaces
            "   ",                  // triple whitespace
            "	",              // &#09;
            "\n\n\r\t   	",
            "  \n \n\r   \t  	 ");
        
    List<String> stripWhitespaces = whitespaces.stream()
            .map(str -> str.replaceAll("\\s+", "' '")).toList();
    
    System.out.println(stripWhitespaces);

Output:

    [' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ']

Tibor17 commented:
@dsubelman 
Why you use quotation marks `' '`?
These are Linux/Unix specific?
We cannot use it if it is platform specific. We have the library `maven-shared-utils` with CLI utility which applies platform specific escaping characters.

Tibor17 commented:
@mthmulders 

> > @mthmulders I have one more question. The developers would not be able to see the TAB because it is a white space. If the new guy would refactor it again, then this IT would not cover the purpose that it has now. I know it is a question of how the xml parser interprets the escaping characters in Maven but I would like to ask you for testing with escaping characters `&#09;` on Maven 3.2.x and 3.8.x. WDYT?
> 
> That's why I added XML comments saying those lines must be tab-indented. I've just tested locally: an XML file with `&#09;` also breaks before the change in `DefaultForkConfiguration` (tested with Maven 4 SNAPSHOT, Maven 3.8.5 and Maven 3.2.5). So it seems to work equally fine for the test. Just tell me which one you prefer @Tibor17: tab characters, XML entities, or both.

As you said, the TAB as a white space works equaly with `<argLine>-Dfoo.property="foo foo/foo/bar/1.0" &#09; -Dbar.property="bar bar/foo/bar/2.0"</argLine>`. AFAIK the developers they press CTRL+ALT+L in IntelliJ IDEA, sometimes they are blind to see all the comments and therefore nothing stops them to refactor the the source code. If they do it, then the character `&#09;` would stay forever, so it would be the strongest guarantee that the purpose is still there.

dsubelman commented:
> @dsubelman Why you use quotation marks `' '`? These are Linux/Unix specific? We cannot use it if it is platform specific. We have the library `maven-shared-utils` with CLI utility which applies platform specific escaping characters. So, it is enough if we keep using one space character as before.

It was to show the difference between the two approaches. Otherwise, you wouldn't be able to see all the difference between the outputs. Instead, I would use my initial proposal:

    argLine.replaceAll("\\s+", " ");

Tibor17 commented:
@dsubelman `argLine.replaceAll("\\s+", " ");` is nice.
The code we have now still cannot trust the values which have a new line in system properties, e.g. `-DmyProp=my \n prop` but we cannot do anyting about it because we do not have smart mechanism to understand the structure of `argLine`. Till now nobody had any objections. The solution that we have is a kind of "too technical" but maybe this would stay untill the people would shout too much.

dsubelman commented:
> @dsubelman `argLine.replaceAll("\\s+", " ");` is nice. The code we have now still cannot trust the values which have a new line in system properties, e.g. `-DmyProp=my \n prop` but we cannot do anyting about it because we do not have smart mechanism to understand the structure of `argLine`. Till now nobody had any objections. The solution that we have is a kind of "too technical" but maybe this would stay untill the people would shout too much.

When you say '_maybe this would stay until the people would shout too much_', do you mean to merge this pull request or not?

Tibor17 commented:
> > @dsubelman `argLine.replaceAll("\\s+", " ");` is nice. The code we have now still cannot trust the values which have a new line in system properties, e.g. `-DmyProp=my \n prop` but we cannot do anyting about it because we do not have smart mechanism to understand the structure of `argLine`. Till now nobody had any objections. The solution that we have is a kind of "too technical" but maybe this would stay untill the people would shout too much.
> 
> When you say '_maybe this would stay until the people would shout too much_', do you mean to merge this pull request or not?

yes, of course, I would be glad to do that.
I only want to open a discussion and tell you what I am thinking about and I want your opinions to be heard.

dsubelman commented:
> > > @dsubelman `argLine.replaceAll("\\s+", " ");` is nice. The code we have now still cannot trust the values which have a new line in system properties, e.g. `-DmyProp=my \n prop` but we cannot do anyting about it because we do not have smart mechanism to understand the structure of `argLine`. Till now nobody had any objections. The solution that we have is a kind of "too technical" but maybe this would stay untill the people would shout too much.
> > 
> > 
> > When you say '_maybe this would stay until the people would shout too much_', do you mean to merge this pull request or not?
> 
> yes, of course, I would be glad to do that. I only want to open a discussion and tell you what I am thinking about and I want your opinions to be heard.

Understood, thanks for the clarification. 

Why would you use `\n` in system properties? In those cases, if you replace the `\n` with whitespace, change the behavior? In other words, you may need not remove `\n` in those cases?

Tibor17 commented:
> > > > @dsubelman `argLine.replaceAll("\\s+", " ");` is nice. The code we have now still cannot trust the values which have a new line in system properties, e.g. `-DmyProp=my \n prop` but we cannot do anyting about it because we do not have smart mechanism to understand the structure of `argLine`. Till now nobody had any objections. The solution that we have is a kind of "too technical" but maybe this would stay untill the people would shout too much.
> > > 
> > > 
> > > When you say '_maybe this would stay until the people would shout too much_', do you mean to merge this pull request or not?
> > 
> > 
> > yes, of course, I would be glad to do that. I only want to open a discussion and tell you what I am thinking about and I want your opinions to be heard.
> 
> Understood, thanks for the clarification.
> 
> Why would you need to use `\n` in system properties? (like in your example `-DmyProp=my \n prop`) In those cases, if you replace the `\n` with whitespace, change the behavior? In other words, you may need not remove `\n` in those cases?

@dsubelman 
I don't need `\n` :-) but the users sometimes have crazy requirements. I just wanted to be prepared with some thoughts, nothing else. ok, we can finish this PR.

@mthmulders 
WDYT about using `argLine.replaceAll("\\s+", " ");` instead of what we have today?

mthmulders commented:
> @mthmulders
> WDYT about using `argLine.replaceAll("\\s+", " ");` instead of what we have today?

I like it, replaced the original implementation with this one.

> If they do it, then the character `&#09;` would stay forever, so it would be the strongest guarantee that the purpose is still there.

Makes sense, I've replaced the tab chars in the IT with XML entities.

Tibor17 commented on surefire-its/src/test/resources/argLine-parameter/pom.xml:
```diff
@@ -52,7 +52,14 @@
+            regular tab characters could easily be reformatted by an IDE,
+            so we're using XML entities here.
+            -->
+            -Dfoo.property="foo foo/foo/bar/1.0";&#09;-Dbar.property="bar bar/foo/bar/2.0"
```
@mthmulders 
Why the special character `&#09;` starts with semicollon, like this `;&#09;`?

mthmulders replied:
A mistake on my side. I'll fix it.

Tibor17 commented:
@mthmulders LGTM

dsubelman commented:
@mthmulders, @Tibor17, thanks for fixing this issue.

jira-importer commented:
Resolve #3112
